File: applications/benchmark_model/python/benchmark_image_model.py
# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os, sys
from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter

from onnx_helper import thread_cuda_init, thread_cuda_deinit, ONNXClassifierWrapper
from PIL import Image
import numpy as np
import cupy as cp

from holoscan.conditions import CountCondition
from holoscan.core import Application
from holoscan.core import (
    ExecutionContext,
    InputContext,
    Operator,
    OperatorSpec,
    OutputContext,
    Tensor,
)
from holoscan.operators import (
    FormatConverterOp,
    HolovizOp,
    InferenceOp,
    SegmentationPostprocessorOp,
    VideoStreamReplayerOp,
)
from holoscan.resources import UnboundedAllocator
from holoscan.gxf import Entity
import holoscan as hs

myfolder = "../images/"


def main(args):
    app = App(
        args.data,
        args.model_name,
        args.image_folder,
        args.multi_inference,
        args.only_inference,
        args.inference_postprocessing,
    )
    # if the --config command line argument was provided, it will override this config_file
    app.config(config_file)
    app.run()


class ReadImagesOp(Operator):
    def __init__(self, fragment, *args, **kwargs):
        super().__init__(fragment, *args, **kwargs)
        self.ctx = None

    def setup(self, spec: OperatorSpec):
        spec.output("out")

    def start(self):
        if not os.path.exists(myfolder):
            print("folder does not exist")
            return False
        self.filenames = os.listdir(myfolder)
        self.index = 0

    def stop(self):
        # if self.ctx:
        #     thread_cuda_deinit(self.ctx)
        pass

    def compute(self, op_input: InputContext, op_output: OutputContext, context: ExecutionContext):
        PRECISION = np.float32

        image_path = os.path.join(
            myfolder, self.filenames[self.index]
        )  # Change this to the path of your image
        if not os.path.exists(image_path):
            print("image path does not exist")
            return
        self.index += 1
        print(image_path)
        image = Image.open(image_path).convert("RGB")

        # Apply preprocessing
        resize = 256
        crop_size = 224

        # Resize
        image = image.resize((resize, resize))

        # Center crop
        left = (resize - crop_size) // 2
        top = (resize - crop_size) // 2
        right = left + crop_size
        bottom = top + crop_size
        image = image.crop((left, top, right, bottom))

        image = np.array(image, dtype=PRECISION) / 255.0  # Normalize the pixel values to [0, 1]
        print(image.shape)

        # Transpose to match ONNX model input format (B, C, H, W)
        image = np.transpose(image, (2, 0, 1))
        print(image.shape)
        image = np.expand_dims(image, axis=0)  # Add batch dimension
        print(image.shape)
        # image = np.ascontiguousarray(image.copy())

        # image = np.moveaxis(image, 2, 0)[None]
        imagetensor = cp.asarray(image)
        imagetensor = Tensor.as_tensor(imagetensor)

        print("Shape:", imagetensor.shape)
        out_message = {}
        out_message["source_image"] = imagetensor

        op_output.emit(out_message, "out")
        # op_output.emit(image, "out")


class ImageClassificationOp(Operator):
    def __init__(self, fragment, *args, **kwargs):
        super().__init__(fragment, *args, **kwargs)
        self.ctx = None

    def setup(self, spec: OperatorSpec):
        spec.input("in")
        spec.output("out")

    def start(self):
        self.ctx = thread_cuda_init()

    def stop(self):
        if self.ctx:
            thread_cuda_deinit(self.ctx)

    def compute(self, op_input: InputContext, op_output: OutputContext, context: ExecutionContext):
        image = op_input.receive("in")
        # print(image)
        BATCH_SIZE = 1  # We process one image at a time
        PRECISION = np.float32
        N_CLASSES = 1000  # Our ResNet-50 is trained on a 1000 class ImageNet task

        # Load and preprocess the image
        trt_model = ONNXClassifierWrapper(
            "../data/resnet50/resnet_engine.trt",
            [BATCH_SIZE, N_CLASSES],
            target_dtype=PRECISION,
        )
        # Make predictions
        predictions = trt_model.predict(image)

        op_output.emit(predictions, "out")


class PrintTextOp(Operator):
    def __init__(self, fragment, *args, **kwargs):
        super().__init__(fragment, *args, **kwargs)

    def setup(self, spec: OperatorSpec):
        spec.input("in")

    def start(self):
        with open("../data/imagenet_classes.txt", "r") as f:
            self.labels = f.read().splitlines()
            self.labels = [l.strip() for l in self.labels]

    def stop(self):
        pass

    def compute(self, op_input: InputContext, op_output: OutputContext, context: ExecutionContext):
        predictions = op_input.receive("in").get("output")
        predictions = cp.asarray(predictions).get()
        # Get the top 5 probability scores and class labels
        top_5_indices = np.argsort(predictions[0])[::-1][:5]
        print(top_5_indices)

        # Print top 5 probability scores and class labels
        print("Top 5 Predictions:")
        for idx in top_5_indices:
            print(f"Class {idx}: {self.labels[idx]} - Probability {predictions[0][idx]*100:.2f}%")


class App(Application):
    def __init__(
        self,
        datapath,
        model_name,
        image_folder,
        num_inferences,
        only_inference,
        inference_postprocessing,
    ):
        """Initialize the application"""

        super().__init__()

        self.name = "Benchmark Image Model App"

        self.datapath = datapath
        self.model_name = model_name
        self.image_folder = image_folder
        self.num_inferences = num_inferences
        self.only_inference = only_inference
        self.inference_postprocessing = inference_postprocessing

        if not os.path.exists(self.datapath):
            raise ValueError(f"Data path {self.datapath} does not exist.")

        self.model_path = os.path.join(self.datapath, model_name)
        if not os.path.exists(self.model_path):
            raise ValueError(f"Model path {self.model_path} does not exist.")

        if not os.path.exists(self.image_folder):
            raise ValueError(f"Image folder {self.image_folder} does not exist.")

    def compose(self):
        host_allocator = UnboundedAllocator(self, name="host_allocator")

        readimages = ReadImagesOp(self, CountCondition(self, 3), name="source")

        # model_path_map = {"own_model": "../data/resnet50/resnet_engine.trt"}
        model_path_map = {"own_model": "../data/resnet50/model.onnx"}
        pre_processor_map = {"own_model": ["source_image"]}
        inference_map = {"own_model": ["output"]}
        inference = InferenceOp(
            self,
            name="inference",
            allocator=host_allocator,
            model_path_map=model_path_map,
            pre_processor_map=pre_processor_map,
            inference_map=inference_map,
            # is_engine_path=True,
            **self.kwargs("inference"),
        )
        # inference = ImageClassificationOp(self, name="inference")

        textoutput = PrintTextOp(self, name="textoutput")

        self.add_flow(readimages, inference, {("", "receivers")})
        self.add_flow(inference, textoutput, {("transmitter", "in")})
        # self.add_flow(readimages, inference)
        # self.add_flow(inference, textoutput, {("out", "in")})

        # attach count condition to inference


if __name__ == "__main__":
    # Parse args
    parser = ArgumentParser(
        description="Benchmark Model Application.",
        formatter_class=ArgumentDefaultsHelpFormatter,
    )
    default_data_path = os.environ.get("HOLOSCAN_INPUT_PATH", "../data")
    parser.add_argument(
        "-d",
        "--data",
        type=str,
        default=default_data_path,
        help="Path to the data directory",
    )
    parser.add_argument(
        "-m",
        "--model-name",
        type=str,
        default="identity_model.onnx",
        help="Path to the model directory",
    )
    parser.add_argument(
        "-a", "--image-folder", type=str, default="images", help="Path to the image folder"
    )
    parser.add_argument(
        "-i",
        "--only-inference",
        action="store_true",
        help="Only run inference, no post-processing or visualization",
    )
    parser.add_argument(
        "-p",
        "--inference-postprocessing",
        action="store_true",
        help="Run inference and post-processing, no visualization",
    )
    parser.add_argument(
        "-l",
        "--multi-inference",
        type=int,
        default=1,
        help="Number of inferences to run in parallel",
    )
    # add positional argument CONFIG which is just a string
    config_file = os.path.join(os.path.dirname(__file__), "benchmark_model.yaml")
    parser.add_argument("ConfigPath", nargs="?", default=config_file, help="Path to config file")

    args = parser.parse_args()
    main(args)

File: applications/benchmark_model/cpp/benchmark_image_model.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <getopt.h>

#include <holoscan/operators/format_converter/format_converter.hpp>
#include <holoscan/operators/holoviz/holoviz.hpp>
#include <holoscan/operators/inference/inference.hpp>
#include <holoscan/operators/segmentation_postprocessor/segmentation_postprocessor.hpp>
#include <holoscan/operators/video_stream_replayer/video_stream_replayer.hpp>
#include "holoscan/holoscan.hpp"

#include <opencv2/opencv.hpp>

#include "cvcuda_utils.hpp"
#include "gxf_utils.hpp"

#include <NvInfer.h>
#include <cuda_runtime.h>
#include <cuda_runtime_api.h>
#include <nvcv/NvDecoder.h>
#include <cvcuda/OpConvertTo.hpp>
#include <cvcuda/OpCustomCrop.hpp>
#include <cvcuda/OpNormalize.hpp>
#include <cvcuda/OpReformat.hpp>
#include <cvcuda/OpResize.hpp>
#include <nvcv/Image.hpp>
#include <nvcv/Tensor.hpp>
#include <nvcv/TensorDataAccess.hpp>

#define N_CLASSES 1000

namespace holoscan::ops {

/**
 * @brief This Operator takes an input tensor and does nothing with it. It works as a sink
 * without any function.
 *
 */
class SinkOp : public holoscan::Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(SinkOp)

  SinkOp() = default;

  void setup(OperatorSpec& spec) { spec.input<std::any>("in"); }

  void compute(InputContext& op_input, OutputContext& op_output, ExecutionContext&) {
    // auto value = op_input.receive<std::any>("in");
    auto value = op_input.receive<holoscan::TensorMap>("in");
    auto maybe_tensormap = value.value();
    // iterate key value of this map, printing all keys
    for (auto& [key, tensor] : maybe_tensormap) { std::cout << "Key: " << key << std::endl; }
  }
};

class ReadImagesOp : public holoscan::Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(ReadImagesOp)

  ReadImagesOp() = default;

  void setup(OperatorSpec& spec) { spec.output<holoscan::TensorMap>("image_tensor"); }

  void start() override {
    // Check if the directory exists
    if (!std::filesystem::exists(image_directory)) {
      std::cerr << "Image directory " << image_directory << " does not exist." << std::endl;
      return;
    }
    for (const auto& entry : std::filesystem::directory_iterator(image_directory)) {
      files.push_back(entry.path());
    }
    cudaStreamCreate(&cudastream);
  }

  void stop() override { cudaStreamDestroy(cudastream); }

  void compute(InputContext& op_input, OutputContext& op_output, ExecutionContext& context) {
    using namespace cv;
    // list all the files in the directory
    std::string current_file = files[count++];

    HOLOSCAN_LOG_INFO("Current file {}", current_file);

    Mat image = imread(current_file, IMREAD_COLOR);
    if (image.empty()) {
      std::cout << "Could not open or find the image" << std::endl;
      return;
    }

    // Convert the image to RGB8 array
    // OpenCV stores images in BGR format, so we need to convert it to RGB
    Mat rgbImage;
    cvtColor(image, rgbImage, COLOR_BGR2RGB);

    // Access the RGB8 array
    uchar* rgbArray = rgbImage.data;
    int width = rgbImage.cols;
    int height = rgbImage.rows;

    uint32_t batchSize = 1;
    int totalImages = 1;
    int maxImageWidth = width;
    int maxImageHeight = height;
    int maxChannels = 3;

    nvcv::TensorDataStridedCuda::Buffer inBuf;
    inBuf.strides[3] = sizeof(uint8_t);
    inBuf.strides[2] = maxChannels * inBuf.strides[3];
    inBuf.strides[1] = maxImageWidth * inBuf.strides[2];
    inBuf.strides[0] = maxImageHeight * inBuf.strides[1];
    cudaMallocAsync(&inBuf.basePtr, batchSize * inBuf.strides[0], cudastream);

    nvcv::Tensor::Requirements inReqs =
        nvcv::Tensor::CalcRequirements(batchSize, {width, height}, nvcv::FMT_RGB8);

    nvcv::TensorDataStridedCuda inData(nvcv::TensorShape{inReqs.shape, inReqs.rank, inReqs.layout},
                                       nvcv::DataType{inReqs.dtype},
                                       inBuf);

    nvcv::Tensor inTensor = TensorWrapData(inData);

    // NvJpeg is used to load the images to create a batched input device buffer.
    uint8_t* gpuInput = reinterpret_cast<uint8_t*>(inBuf.basePtr);
    cudaMemcpy(inBuf.basePtr, rgbArray, width * height * 3 * sizeof(uchar), cudaMemcpyHostToDevice);

    // Apply preprocessing
    int resize = 256;

    nvcv::Tensor::Requirements outReqs =
        nvcv::Tensor::CalcRequirements(batchSize, {resize, resize}, nvcv::FMT_RGB8);

    nvcv::TensorDataStridedCuda::Buffer outBuf;
    std::copy(outReqs.strides, outReqs.strides + NVCV_TENSOR_MAX_RANK, outBuf.strides);

    int64_t outLayerSize = nvcv::CalcTotalSizeBytes(nvcv::Requirements{outReqs.mem}.cudaMem());
    cudaMalloc(&outBuf.basePtr, outLayerSize);

    nvcv::TensorDataStridedCuda outData(
        nvcv::TensorShape{outReqs.shape, outReqs.rank, outReqs.layout},
        nvcv::DataType{outReqs.dtype},
        outBuf);
    nvcv::Tensor outTensor = TensorWrapData(outData);

    // Apply resize
    cvcuda::Resize resizeOp;
    HOLOSCAN_LOG_INFO("Will resize");
    resizeOp(cudastream, inTensor, outTensor, NVCV_INTERP_NEAREST);

    HOLOSCAN_LOG_INFO("Resize succesful");

    int crop_size = 224;
    int xy_start = (resize - crop_size) / 2;

    NVCVRectI crpRect = {xy_start, xy_start, crop_size, crop_size};

    nvcv::Tensor cropTensor(batchSize, {crop_size, crop_size}, nvcv::FMT_RGB8);
    cvcuda::CustomCrop cropOp;
    cropOp(cudastream, outTensor, cropTensor, crpRect);

    // Convert to FP32
    nvcv::Tensor floatTensor(batchSize, {resize, resize}, nvcv::FMT_RGBf32);
    cvcuda::ConvertTo convertOp;
    convertOp(cudastream, cropTensor, floatTensor, 1.0f / 255.f, 0.0f);

#if 0
    // Normalize the pixel values
    nvcv::Tensor::Requirements reqsScale =
        nvcv::Tensor::CalcRequirements(1, {1, 1}, nvcv::FMT_RGBf32);
    int64_t scaleBufferSize = CalcTotalSizeBytes(nvcv::Requirements{reqsScale.mem}.cudaMem());
    nvcv::TensorDataStridedCuda::Buffer bufScale;
    std::copy(reqsScale.strides, reqsScale.strides + NVCV_TENSOR_MAX_RANK, bufScale.strides);
    cudaMalloc(&bufScale.basePtr, scaleBufferSize);
    nvcv::TensorDataStridedCuda scaleIn(
        nvcv::TensorShape{reqsScale.shape, reqsScale.rank, reqsScale.layout},
        nvcv::DataType{reqsScale.dtype},
        bufScale);
    nvcv::Tensor stddevTensor = TensorWrapData(scaleIn);

    nvcv::TensorDataStridedCuda::Buffer bufBase;
    nvcv::Tensor::Requirements reqsBase =
        nvcv::Tensor::CalcRequirements(1, {1, 1}, nvcv::FMT_RGBf32);
    int64_t baseBufferSize = CalcTotalSizeBytes(nvcv::Requirements{reqsBase.mem}.cudaMem());
    std::copy(reqsBase.strides, reqsBase.strides + NVCV_TENSOR_MAX_RANK, bufBase.strides);
    cudaMalloc(&bufBase.basePtr, baseBufferSize);
    nvcv::TensorDataStridedCuda baseIn(
        nvcv::TensorShape{reqsBase.shape, reqsBase.rank, reqsBase.layout},
        nvcv::DataType{reqsBase.dtype},
        bufBase);
    nvcv::Tensor meanTensor = TensorWrapData(baseIn);

    float stddev[3] = {0.229, 0.224, 0.225};
    float mean[3] = {0.485f, 0.456f, 0.406f};
    auto meanData = meanTensor.exportData<nvcv::TensorDataStridedCuda>();
    auto stddevData = stddevTensor.exportData<nvcv::TensorDataStridedCuda>();

    uint32_t flags = CVCUDA_NORMALIZE_SCALE_IS_STDDEV;
    cudaMemcpyAsync(
        stddevData->basePtr(), stddev, 3 * sizeof(float), cudaMemcpyHostToDevice, cudastream);
    cudaMemcpyAsync(
        meanData->basePtr(), mean, 3 * sizeof(float), cudaMemcpyHostToDevice, cudastream);

    nvcv::Tensor normTensor(batchSize, {crop_size, crop_size}, nvcv::FMT_RGBf32);

    cvcuda::Normalize normalizeOp;
    normalizeOp(
        cudastream, cropTensor, meanTensor, stddevTensor, normTensor, 1.0f, 0.0f, 0.0f, flags);
#endif

    nvcv::Tensor::Requirements reqsOutputLayer =
        nvcv::Tensor::CalcRequirements(batchSize, {crop_size, crop_size}, nvcv::FMT_RGBf32p);

    nvcv::Tensor outputLayerTensor(
        nvcv::TensorShape{reqsOutputLayer.shape, reqsOutputLayer.rank, reqsOutputLayer.layout},
        nvcv::DataType{reqsOutputLayer.dtype});

    const auto& [out_message, tensor_data_pointer] =
        create_out_message_with_tensor_like(context.context(), outputLayerTensor);

    nvcv::TensorDataStridedCuda::Buffer cv_out_buffer;
    std::copy(reqsOutputLayer.strides,
              reqsOutputLayer.strides + NVCV_TENSOR_MAX_RANK,
              cv_out_buffer.strides);
    cv_out_buffer.basePtr = static_cast<NVCVByte*>(*tensor_data_pointer);
    nvcv::TensorDataStridedCuda out_data(
        nvcv::TensorShape{reqsOutputLayer.shape, reqsOutputLayer.rank, reqsOutputLayer.layout},
        nvcv::DataType{reqsOutputLayer.dtype},
        cv_out_buffer);
    nvcv::Tensor cv_out_tensor = nvcv::TensorWrapData(out_data);

    cvcuda::Reformat reformatOp;
    // reformatOp(cudastream, normTensor, cv_out_tensor);
    // reformatOp(cudastream, cropTensor, cv_out_tensor);
    reformatOp(cudastream, floatTensor, cv_out_tensor);

    std::cout << "cv_out_tensor shape: " << cv_out_tensor.shape() << std::endl;

    cudaStreamSynchronize(cudastream);

    op_output.emit(out_message, "image_tensor");
  }

 private:
  std::string image_directory = "../images/";
  std::vector<std::string> files;
  int count = 0;
  // read the image from the current_file using cvcuda
  cudaStream_t cudastream;
};

class PrintTextOp : public holoscan::Operator {
 public:
  HOLOSCAN_OPERATOR_FORWARD_ARGS(PrintTextOp)

  PrintTextOp() = default;

  void start() override {
    std::ifstream file(image_classes_file);
    std::string line;
    while (std::getline(file, line)) { labels.push_back(line); }
  }

  void setup(OperatorSpec& spec) { spec.input<holoscan::TensorMap>("in"); }

  void compute(InputContext& op_input, OutputContext& op_output, ExecutionContext&) {
    auto value = op_input.receive<holoscan::TensorMap>("in");
    auto maybe_tensor = value.value().at("output");

    auto predictions = maybe_tensor.get()->data();
    auto bytes = maybe_tensor.get()->nbytes();

    std::cout << "bytes: " << bytes << std::endl;
    // allocate host memory
    float* host_predictions = (float*)malloc(bytes);
    cudaMemcpy(host_predictions, predictions, bytes, cudaMemcpyDeviceToHost);

    // sync default stream
    cudaStreamSynchronize(0);

    HOLOSCAN_LOG_DEBUG("Memcpy Successful");

    // convert host_predictions[0]...[N_CLASSES] to a vector
    std::vector<float> predictions_vec(host_predictions, host_predictions + N_CLASSES);

    HOLOSCAN_LOG_DEBUG("Vector created");

    // sort the vector and return the top 5 indices
    std::vector<int> indices(N_CLASSES);
    std::iota(indices.begin(), indices.end(), 0);
    std::partial_sort(
        indices.begin(), indices.begin() + 5, indices.end(), [&predictions_vec](int i1, int i2) {
          return predictions_vec[i1] > predictions_vec[i2];
        });

    HOLOSCAN_LOG_INFO("Top 5 predictions: ");
    for (int i = 0; i < 5; i++) {
      // .2f decimal points
      std::cout << "Class " << indices[i] << ": " << labels[indices[i]]
                << " - Probability: " << std::fixed << std::setprecision(2)
                << predictions_vec[indices[i]] * 100 << "%" << std::endl;
    }
  }

 private:
  std::string image_classes_file = "../data/imagenet_classes.txt";
  std::vector<std::string> labels;
};
}  // namespace holoscan::ops

class App : public holoscan::Application {
 public:
  App() { holoscan::Application(); }

  void compose() override {
    using namespace holoscan;

    std::shared_ptr<Resource> pool_resource = make_resource<UnboundedAllocator>("pool");
    // Create the operators
    auto read_images_op = make_operator<holoscan::ops::ReadImagesOp>(
        "read_images_op", make_condition<holoscan::CountCondition>(3));
    auto sink_op = make_operator<holoscan::ops::SinkOp>("sink_op");

    ops::InferenceOp::DataMap model_path_map;
    ops::InferenceOp::DataVecMap pre_processor_map;
    ops::InferenceOp::DataVecMap inference_map;

    std::string model_index_str = "own_model";

    model_path_map.insert(model_index_str, "../data/resnet50/model.onnx");
    pre_processor_map.insert(model_index_str, {"image"});

    std::string output_name = "output";
    inference_map.insert(model_index_str, {output_name});

    auto inference = make_operator<ops::InferenceOp>("inference",
                                                     from_config("inference"),
                                                     Arg("allocator") = pool_resource,
                                                     Arg("model_path_map", model_path_map),
                                                     Arg("pre_processor_map", pre_processor_map),
                                                     Arg("inference_map", inference_map));

    // Connect the operators
    // add_flow(read_images_op, sink_op);
    add_flow(read_images_op, inference, {{"", "receivers"}});

    auto print_text_op = make_operator<ops::PrintTextOp>("print_text_op");
    add_flow(inference, print_text_op);
  }

 private:
  int num_inferences = 1;
  bool only_inference = false, inference_postprocessing = false;
  std::string datapath, model_name, video_name;
};

int main(int argc, char** argv) {
  auto app = holoscan::make_application<App>();
  std::string config_name = "";
  if (config_name != "") {
    // Check if config_name is a valid path
    if (!std::filesystem::exists(config_name)) {
      std::cerr << "Config file " << config_name << " does not exist." << std::endl;
      return 0;
    }
    app->config(config_name);
  } else {
    auto config_path = std::filesystem::canonical(argv[0]).parent_path();
    config_path += "/benchmark_model.yaml";
    app->config(config_path);
  }

  app->run();
  return 0;
}